### PR TITLE
Added offset to the token location

### PR DESCRIPTION
**Pull request by AsterMiha**
I added a field to the Loc structure but to address the concerns raised [here](https://github.com/dlang/dmd/pull/11788#issuecomment-701545320) these changes should only be available when building for a certain version. At the moment, the version part is commented so the lexer unittest I added could run.
How can I make this test run only when building with the correct version?

### CODE REVIEW

**Comment by MoonlightSentinel** [comment]
> How can I make this test run only when building with the correct version?

You can simply add the version here (as long as you are only using `unit` tests):

https://github.com/dlang/dmd/blob/5fda4ea088f70afb99818482e8482771758136a2/test/tools/unit_test_runner.d#L246-L260

File: src/dmd/globals.d [hunk 1]
```diff
@@ -572,6 +574,15 @@ nothrow:
         this.filename = filename;
     }
 
+    //version(ExtraLocation)
+        extern (D) this(const(char)* filename, uint linnum, uint charnum, uint offset) pure
```
Comment (MoonlightSentinel) [inline src/dmd/globals.d:578]: It's probably better to simply add the parameter by default and discard the value without `version(ExtraLocation`)

**Comment by RazvanN7** [comment]
cc @jacob-carlborg How does this look?

**Comment by RazvanN7** [comment]
One comment that I have is that we should unify all changes under a common switch called "DMDLIB". Therefore, replace the EXTRA_LOCATION switch with DMDLIB and maybe in a subsequent PR replace all other variations on this topic ([1][2])

[1]  https://github.com/dlang/dmd/pull/11092
[2] https://github.com/dlang/dmd/pull/11265

**Comment by thewilsonator** [comment]
The whole point of using offsets was to *reduce* the size of `Loc` and therefore the memory the compiler consumes.
This could be done by an indirection/offset table to map `(size_t (or unit) x Table)  -> (filename, linenum, charnum)`. 
How this mapping is done is up to you but it should be modular and I'd like to see the numbers before this is merged.

**Comment by RazvanN7** [comment]
@thewilsonator That is a different point. This change is trying to add offset location for tools that use dmd as a library. As you can see it is hided by a compiler version. The refactoring that you are talking about may be the subject of a different PR and indeed might help saving some space, but the purpose of  this PR is entirely different.

**Comment by RazvanN7** [comment]
Actually, thinking more about it, this does seem like a great opportunity for optimization.

File: src/dmd/globals.d [hunk 2]
```diff
@@ -561,15 +561,19 @@ struct Loc
     const(char)* filename; // either absolute or relative to cwd
     uint linnum;
     uint charnum;
+    version(ExtraLocation)
```
Comment (jacob-carlborg) [inline src/dmd/globals.d:564]: Space after `version`.

File: src/dmd/globals.d [hunk 3]
```diff
@@ -561,15 +561,19 @@ struct Loc
...
     static immutable Loc initial;       /// use for default initialization of const ref Loc's
 
 nothrow:
-    extern (D) this(const(char)* filename, uint linnum, uint charnum) pure
+    extern (D) this(const(char)* filename, uint linnum, uint charnum, uint offset = 0) pure
     {
         this.linnum = linnum;
         this.charnum = charnum;
         this.filename = filename;
+        version(ExtraLocation)
```
Comment (jacob-carlborg) [inline src/dmd/globals.d:575]: Same here, space after `version`.

File: src/dmd/globals.d [hunk 4]
```diff
@@ -561,15 +561,19 @@ struct Loc
...
     uint linnum;
     uint charnum;
+    version(ExtraLocation)
+        uint offset;
 
     static immutable Loc initial;       /// use for default initialization of const ref Loc's
 
 nothrow:
-    extern (D) this(const(char)* filename, uint linnum, uint charnum) pure
+    extern (D) this(const(char)* filename, uint linnum, uint charnum, uint offset = 0) pure
```
Comment (jacob-carlborg) [inline src/dmd/globals.d:570]: I don't mind, but Walter doesn't like parameters with default values. Is there a point to have this parameter since it's not used.

File: test/tools/unit_test_runner.d [hunk 5]
```diff
@@ -248,6 +248,7 @@ void writeCmdfile(string path, string runnerPath, string outputPath,
         "-version=GC",
         "-version=NoMain",
         "-version=MARS",
+        "-version=ExtraLocation",
```
Comment (jacob-carlborg) [inline test/tools/unit_test_runner.d:251]: This needs to be added to the Dub file as well. To the lexer subpackage: https://github.com/dlang/dmd/blob/37556c0c3f7aca3e37db02631ef2e2c9ffb8a57b/dub.sdl#L16

Otherwise, when the Dub package is used as a library, it won't have this version identifier enabled.

File: test/unit/location.d [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    ];
+
+    foreach(exp; expected)
+    {
+        lex1.nextToken();
+        assert(lex1.token.loc.linnum == exp[0]);
+        assert(lex1.token.loc.charnum == exp[1]);
+        assert(lex1.token.loc.offset == exp[2]);
+    }
+}
```
Comment (jacob-carlborg) [inline test/unit/location.d:35]: Missing newline at the end.

File: test/unit/location.d [hunk 7]
```diff
@@ -0,0 +1,35 @@
+module location;
+
+import dmd.lexer;
+import dmd.globals : Loc;
+
+unittest
```
Comment (jacob-carlborg) [inline test/unit/location.d:6]: All unit test blocks in the `test/unit` directory should have one or multiple UDAs attached to the unit test block containing a string which describes what the test is testing. This does three things:

* Adds a description of what the test is actually testing
* The unit test runner will output this description if the test failed
* It's possible to filter the tests based on this description. This allows to select individual unit test blocks to be executed, instead of having to execute all tests in a file or the whole test suite

File: test/unit/location.d [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+{
+    string text = "int a b xyz\nfloat\n\nchar a=b + c[11];";
```
Comment (jacob-carlborg) [inline test/unit/location.d:8]: I expect all tokens to be tested, not just a few. I also expect edge cases to be test, like:

* The first token in the source code
* The last token (i.e. last byte) in the source code
* Test the location of a token after Windows line endings
* Test the location of a token after Mac line endings (if DMD supports those)
* Test the location of a token after Unix line endings
* The end of file token
* A token after the `#line` token
* A token after a comment (all variants of comments)
* Characters and string tokens with multibyte characters

Here's an example of how I've implemented tests for all tokens:

https://github.com/jacob-carlborg/ddc/commit/32d5fe37be211065558f6b44733f059ec9f10c13#diff-ca42ad79bd703d8bea0e34b342881ac9

It's missing some of the above edge cases, but it should provide you with a pretty good starting point.

You can even improve on the above test setup, by instead of storing the tests in an array, store them in an associative array, where the keys are the tokens and the values are the tests. Then you can use loop over the enum members of the `TOK` enum and access the values in the associative array. This will guarantee that if a token is added, it won't get missed in the test because the test will fail.

**Comment by AsterMiha** [comment]
> The whole point of using offsets was to _reduce_ the size of `Loc` and therefore the memory the compiler consumes.
> This could be done by an indirection/offset table to map `(size_t (or unit) x Table) -> (filename, linenum, charnum)`.
> How this mapping is done is up to you but it should be modular and I'd like to see the numbers before this is merged.

Thanks for the suggestion! I will attempt to do this optimization but in a different PR.
Right now I'm working on adding components useful for tools using dmd as a library and since this will be done under a version it shouldn't impact the memory the compiler uses at the moment. So before this refactoring, I will also add the token size (under this version of course).

File: test/unit/location.d [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        [4, 16, 34],
+        [4, 17, 35],
+    ];
+
+    foreach(exp; expected)
+    {
+        lex1.nextToken();
+        assert(lex1.token.loc.linnum == exp[0]);
+        assert(lex1.token.loc.charnum == exp[1]);
+        assert(lex1.token.loc.offset == exp[2]);
```
Comment (jacob-carlborg) [inline test/unit/location.d:33]: I know some will not agree with me, but in my opinion, it's best practice to have one assertion per unit test block. That means one unit test block per test case. That would be one unit test block per token, plus a few more for the edge cases. Again, have a look at my implementation: https://github.com/jacob-carlborg/ddc/commit/32d5fe37be211065558f6b44733f059ec9f10c13#diff-ca42ad79bd703d8bea0e34b342881ac9.

File: test/unit/location.d [hunk 10]
```diff
@@ -0,0 +1,35 @@
+module location;
```
Comment (jacob-carlborg) [inline test/unit/location.d:1]: I recommend placing this in the `lexer` subdirectory. I also recommend renaming this to something like `location_offset`. Because if you're going to add a test for each token, it's going to be quite a lot of code. I would recommend keeping only tests for the location offset in this file and then create separate files other location tests.

File: src/dmd/lexer.d [hunk 11]
```diff
@@ -2284,6 +2284,8 @@ class Lexer
     final Loc loc() pure @nogc
     {
         scanloc.charnum = cast(uint)(1 + p - line);
+        version(ExtraLocation)
+            scanloc.offset = cast(uint)(line - base) + scanloc.charnum - 1;
```
Comment (jacob-carlborg) [inline src/dmd/lexer.d:2288]: I don't think you have to do this complicated calculation. The offset is more or less already stored in the lexer. You can have a look at how it's stored for inside a token.

**Review by jacob-carlborg — CHANGES_REQUESTED** [review]

**Comment by jacob-carlborg** [comment]
> The whole point of using offsets was to reduce the size of Loc and therefore the memory the compiler consumes

@thewilsonator This PR is to store the offset of a token from the start of the source buffer, i.e. byte count. This is required to be able to do source code refactorings.

File: src/dmd/globals.d (hunk 4 above)
Comment (jacob-carlborg) [inline src/dmd/globals.d:570]: I'm wondering if there are other places in the code that needs to be updated as well. I know there are other places in the code which creates locations, not just the lexer.

File: src/dmd/globals.d (hunk 2 above)
Comment (jacob-carlborg) [inline src/dmd/globals.d:564]: I agree with Razvan. We should have a more generic name of the version identifier that works for all additions that are specific to DMD as a library.

File: test/unit/lexer/location_offset.d [hunk 12]
```diff
@@ -0,0 +1,482 @@
...
+    scope lexer = new Lexer("test.d", code.ptr, 0, code.length, 0, 0);
+
+    lexer.nextToken;
+    lexer.nextToken;
+    lexer.nextToken;
+
+    assert(lexer.token.loc.offset == 14, code);
+}
+
+@("last token in the source code")
```
Comment (jacob-carlborg) [inline test/unit/lexer/location_offset.d:47]: These descriptions should ideally be unique within a file. Otherwise it won't be possible to single out a single unit test block and run only that.

File: test/unit/lexer/location_offset.d [hunk 13]
```diff
@@ -0,0 +1,482 @@
...
+
+    lexer.nextToken;
+
+    assert(lexer.token.loc.offset == 14, code);
+}
+
+@("last token in the source code")
+unittest
+{
+    enum code = "token1 token2 token3";
```
Comment (jacob-carlborg) [inline test/unit/lexer/location_offset.d:36]: When I mentioned "last token in the source code", I was more thinking of the offset of the last token would be the same as last byte. Something like `token1 token2 3` and check the offset of `3`.

File: test/unit/lexer/location_offset.d [hunk 14]
```diff
@@ -0,0 +1,482 @@
...
+    lexer.nextToken;
+    lexer.nextToken;
+
+    assert(lexer.token.loc.offset == 5, code);
+}
+
+@("multibyte character token")
+unittest
+{
+    enum code = "🍺";
```
Comment (jacob-carlborg) [inline test/unit/lexer/location_offset.d:138]: This is not a character token. It's probably an identifier token. You need single quotes (`''`) to make it a character literal.

File: test/unit/lexer/location_offset.d [hunk 15]
```diff
@@ -0,0 +1,482 @@
...
+
+    lexer.nextToken;
+
+    assert(lexer.token.loc.offset == 4, code);
+}
+
+@("multibyte string token")
+unittest
+{
+    enum code = "🍺🍺";
```
Comment (jacob-carlborg) [inline test/unit/lexer/location_offset.d:150]: Same here, this is not a string token.

File: test/unit/lexer/location_offset.d [hunk 16]
```diff
@@ -0,0 +1,482 @@
...
+
+    lexer.nextToken;
+
+    assert(lexer.token.loc.offset == 8, code);
+}
+
+@("token after multibyte character token")
+unittest
+{
+    enum code = "🍺 token";
```
Comment (jacob-carlborg) [inline test/unit/lexer/location_offset.d:162]: Not a character token.

File: test/unit/lexer/location_offset.d [hunk 17]
```diff
@@ -0,0 +1,482 @@
...
+    lexer.nextToken;
+    lexer.nextToken;
+
+    assert(lexer.token.loc.offset == 10, code);
+}
+
+@("token after multibyte string token")
+unittest
+{
+    enum code = "🍺🍺 token";
```
Comment (jacob-carlborg) [inline test/unit/lexer/location_offset.d:175]: Not a string token.

File: test/unit/lexer/location_offset.d [hunk 18]
```diff
@@ -135,51 +135,51 @@ unittest
...
-    assert(lexer.token.loc.offset == 4, code);
+    assert(lexer.token.loc.offset == 0, code);
 }
 
-@("multibyte string token")
+@("multibyte character string token")
 unittest
 {
-    enum code = "🍺🍺";
+    enum code = "\"🍺🍺\"";
```
Comment (jacob-carlborg) [inline test/unit/lexer/location_offset.d:150]: For the outer string literal, I recommend using backticks to avoid needing to escape the double quotes.

File: test/unit/lexer/location_offset.d [hunk 19]
```diff
@@ -0,0 +1,482 @@
...
+    TOK.interval : Test("interval"),
+    TOK.voidExpression : Test("voidexp"),
+    TOK.cantExpression : Test("cantexp"),
+    TOK.showCtfeContext : Test("showCtfeContext"),
+
+    TOK.objcClassReference : Test("class"),
+    TOK.vectorArray : Test("vectorArray"),
+];
+
+static foreach (test; tests)
```
Comment (jacob-carlborg) [inline test/unit/lexer/location_offset.d:468]: The point of using an associative array was to iterate the members of the `TOK` enum. Then use each value as a key into the associative array. This will make sure that if a new member is added to `TOK`, this test will fail.

**Comment by jacob-carlborg** [comment]
Looking good, just a few minor things left.

File: test/unit/lexer/location_offset.d [hunk 20]
```diff
@@ -0,0 +1,482 @@
...
+    TOK.vector : Test("__vector"),
+    TOK.pound : Test("pound", "#"),
+
+    TOK.interval : Test("interval"),
+    TOK.voidExpression : Test("voidexp"),
+    TOK.cantExpression : Test("cantexp"),
+    TOK.showCtfeContext : Test("showCtfeContext"),
+
+    TOK.objcClassReference : Test("class"),
+    TOK.vectorArray : Test("vectorArray"),
```
Comment (jacob-carlborg) [inline test/unit/lexer/location_offset.d:465]: You have a couple of tokens here, which do exist in `TOK` but never are produced by the lexer (I don't know why they are in `TOK`), like `TOK.vectorArray`, `TOK.onScopeSuccess` and a couple of others. Please go through each test and remove the ones that are not produced by the lexer.

File: test/unit/lexer/location_offset.d (hunk 19 above)
Comment (AsterMiha) [reply test/unit/lexer/location_offset.d:468]: Something like this:
`static foreach (tok; TOK)`
raises an `invalid foreach aggregate` error
Is there a way to make this work?

Comment (jacob-carlborg) [reply test/unit/lexer/location_offset.d:468]: Yes, it's possible. I'm trying to come up with some good pointers without directly telling you the how the code should look like. I think it will be more beneficial I you manage to find the answer yourself. Should be possible to search for this online.

Comment (AsterMiha) [reply test/unit/lexer/location_offset.d:468]: I did find something about EnumMembers but would it be alright to rely on phobos or is there some equivalent in dmd?

File: src/dmd/globals.d [hunk 21]
```diff
@@ -561,6 +561,8 @@ struct Loc
     const(char)* filename; // either absolute or relative to cwd
     uint linnum;
     uint charnum;
+    version (DMDLIB)
```
Comment (WalterBright) [inline src/dmd/globals.d:564]: The version should be something like LocOffset, because that's what it is. DMDLIB has no particular connection to offset, and should be independent, although it can be set based on DMDLIB.

File: src/dmd/globals.d [hunk 22]
```diff
@@ -561,6 +561,8 @@ struct Loc
...
+        uint offset;
```
Comment (WalterBright) [inline src/dmd/globals.d:565]: `offset` is heavily used in DMD, making it un-greppable. Instead, use fileOffset or whatever it is an offset from.

File: src/dmd/globals.d [hunk 23]
```diff
@@ -555,12 +555,18 @@ alias d_uns32 = uint32_t;
 alias d_int64 = int64_t;
 alias d_uns64 = uint64_t;
 
+version (DMDLIB) {
```
Comment (jacob-carlborg) [inline src/dmd/globals.d:558]: Curly braces on their own lines, or remove.

File: test/unit/lexer/location_offset.d (hunk 19 above)
Comment (jacob-carlborg) [reply test/unit/lexer/location_offset.d]: Should be fine. Alternatively, you can use `__traits(allMembers)` as well [1].

[1] https://dlang.org/spec/traits.html#allMembers.

**Comment by jacob-carlborg** [comment]
Looks good, just the extra tokens left: https://github.com/dlang/dmd/pull/11823#discussion_r505773149.

File: test/unit/lexer/location_offset.d [hunk 24]
```diff
@@ -0,0 +1,489 @@
...
+    "pragma_" : Test("pragma"),
+    "typeid_" : Test("typeid"),
+
+    "lessThan" : Test("less than", "<"),
+    "greaterThan" : Test("greater then", ">"),
+    "lessOrEqual" : Test("less then or equal", "<="),
+    "greaterOrEqual" : Test("greater then or equal", ">="),
+    "equal" : Test("equal", "=="),
+    "notEqual" : Test("not equal", "!="),
+    "is_" : Test("is_"),
```
Comment (jacob-carlborg) [inline test/unit/lexer/location_offset.d:248]: The code should be `is`.

File: test/unit/lexer/location_offset.d [hunk 25]
```diff
@@ -0,0 +1,489 @@
...
+        @(tests[tok].description)
+        unittest
+        {
+            const newCode = "first_token " ~ tests[tok].code;
+
+            scope lexer = new Lexer("test.d", newCode.ptr, 0, newCode.length, 0, 0);
+
+            lexer.nextToken;
+            lexer.nextToken;
+
```
Comment (jacob-carlborg) [inline test/unit/lexer/location_offset.d:485]: A couple of times I have seen that the token and the source code doesn't match. You should probably add an extra assertion (yes, I know that goes agains what I said before about a single assertion per test) that verifies that the input token (that is, the key in the associative array) matches the output token.

File: test/unit/lexer/location_offset.d [hunk 26]
```diff
@@ -0,0 +1,489 @@
...
+    newAnonymousClass,
+    comment,
+    arrayLiteral,
+    assocArrayLiteral,
+    structLiteral,
+    classReference,
+    thrownException,
+    delegatePointer,
+    delegateFunctionPointer,
+    identity,
```
Comment (jacob-carlborg) [inline test/unit/lexer/location_offset.d:442]: This is produced by the lexer, the code for that token is `is`.

File: test/unit/lexer/location_offset.d [hunk 27]
```diff
@@ -0,0 +1,489 @@
...
+    notIdentity,
```
Comment (jacob-carlborg) [inline test/unit/lexer/location_offset.d:443]: This is produced by the lexer, the code for that token is `!is`.

File: test/unit/lexer/location_offset.d [hunk 28]
```diff
@@ -0,0 +1,489 @@
...
+    symbolOffset,
+    variable,
+    arrayLength,
+    dotTemplateDeclaration,
+    declaration,
+    dSymbol,
+    uadd,
+    remove,
+    newAnonymousClass,
+    comment,
```
Comment (jacob-carlborg) [inline test/unit/lexer/location_offset.d:434]: This is produced by the lexer if `commentToken` in the constructor is set to `true`.

File: test/unit/lexer/location_offset.d (hunk 26 above)
Comment (AsterMiha) [reply test/unit/lexer/location_offset.d:481]: The lexer alone seems to only identify the `is_` token. The parser changes this to `identity`.
Same for `!is`. The lexer sees 2 tokens: `not` and `is_`.

Comment (jacob-carlborg) [reply test/unit/lexer/location_offset.d:481]: Oh, I see, my bad.

**Comment by JohanEngelen** [comment]
If you are interested in optimizing source location tracking, I strongly advise reading up on how Clang does that. It has a very efficient implementation. DMD's "Loc" size = 64 + 32 + 32 = 128 bits. Clang's is 32bits.
https://clang.llvm.org/docs/InternalsManual.html#the-sourcelocation-and-sourcemanager-classes

**Comment by AsterMiha** [comment]
@JohanEngelen  Thank you! I will look into it.

**Review by jacob-carlborg — APPROVED** [review]
LGTM.

**Comment by jacob-carlborg** [comment]
@AsterMiha in the future, when fixing something in a PR related to a specific comment. Please replay to that comment that the issue has been fixed and resolve the conversation. This will make it easier for reviewers to see that a particular comment has been addressed.

**Review by RazvanN7 — APPROVED** [review]
